usr/lib/common/pkey_utils.c

```diff
@@ -602,3 +702,343 @@ CK_RV pkey_aes_cmac(OBJECT *key_obj, CK_BYTE *message,
+        DEF_PARAM(P256, 32);
+        DEF_PARAM(P384, 48);
+        DEF_PARAM(P521, 80);
+    } param;
+#undef DEF_PARAM
+
+    CK_RV ret;
+    unsigned long fc;
+    CK_ATTRIBUTE *pkey_attr = NULL;
+    int rc, off;
+    cpacf_curve_type_t curve_type;
+
+    /* Get protected key from key object */
+    if (template_attribute_get_non_empty(privkey->template, CKA_IBM_OPAQUE_PKEY,
+                                         &pkey_attr) != CKR_OK) {
+        TRACE_ERROR("Could not find CKA_IBM_OPAQUE_PKEY in key's template.\n");
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Setup CPACF parmblock */
+    memset(&param, 0, sizeof(param));
+    curve_type = get_cpacf_curve_type(privkey->template);
+    switch (curve_type) {
+    case curve_p256:
+        if (hash_len > 32)
+            hash_len = 32;
+        off = 32 - hash_len;
+        memcpy(param.P256.hash + off, hash, hash_len);
+        memcpy(param.P256.priv, pkey_attr->pValue, 32);
+        memcpy(param.P256.vp, pkey_attr->pValue + 32, 32);
+        *sig_len = 2 * 32;
+        break;
+    case curve_p384:
+        if (hash_len > 48)
+            hash_len = 48;
+        off = 48 - hash_len;
+        memcpy(param.P384.hash + off, hash, hash_len);
+        memcpy(param.P384.priv, pkey_attr->pValue, 48);
+        memcpy(param.P384.vp, pkey_attr->pValue + 48, 32);
+        *sig_len = 2 * 48;
+        break;
+    case curve_p521:
+        if (hash_len > 66)
+            hash_len = 66;
+        /* Note that the pkey for p521 has 80 + 32 bytes. */
+        off = 80 - hash_len;
+        memcpy(param.P521.hash + off, hash, hash_len);
+        memcpy(param.P521.priv, pkey_attr->pValue, pkey_attr->ulValueLen - 32);
+        memcpy(param.P521.vp, pkey_attr->pValue + pkey_attr->ulValueLen - 32, 32);
+        *sig_len = 2 * 66;
+        break;
+    default:
+        TRACE_ERROR("Could not determine the curve type.\n");
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    if (sig == NULL) {
+        /* length_only requested, just return with *sig_len */
+        ret = CKR_OK;
+        goto done;
+    }
+
+    /* Set random number and function code */
+    switch (curve_type) {
+    case curve_p256:
+        fc = KDSA_ENCRYPTED_ECDSA_SIGN_P256 | 0x80;
+        if (rng_cb != NULL) {
+            rng_cb(param.P256.rand, sizeof(param.P256.rand));
+            fc = KDSA_ENCRYPTED_ECDSA_SIGN_P256;
+        }
+        break;
+    case curve_p384:
+        fc = KDSA_ENCRYPTED_ECDSA_SIGN_P384 | 0x80;
+        if (rng_cb != NULL) {
+            rng_cb(param.P384.rand, sizeof(param.P384.rand));
+            fc = KDSA_ENCRYPTED_ECDSA_SIGN_P384;
+        }
+        break;
+    case curve_p521:
+        fc = KDSA_ENCRYPTED_ECDSA_SIGN_P521 | 0x80;
+        if (rng_cb != NULL) {
+            rng_cb(param.P521.rand, sizeof(param.P521.rand));
+            fc = KDSA_ENCRYPTED_ECDSA_SIGN_P521;
+        }
+        break;
+    default:
+        /* Cannot occur */
```

Comment:
If it occurs anyway (e.g., due to a coding bug during addition of a new curve), we will not even notice that we have an unimplemented point causing a sign error.  Even worse, in this case we jump to `done` returning `ret` which is uninitialized so far.  So please initialize `ret` to some error and/or issue a trace call.

Reply:
The curve_type was already checked in the switch above, so we would already be at "done" with a bad rc. There are two "switch (curve_type)" statements in a row. But I will add the trace and rc here again, just for clarity.

Reply:
Sorry, did not see the first switch checked exactly the same thing.  Couldn't we simply get rid of the second switch?  Just set the `fc` in the first switch and remember the two parameters for `rng_cb`.  Then, the second switch boils down to
```
if (rng_cb != NULL) {
    rng_cb(rng_area, rng_area_size);
    fc &= ~0x80u;
}
```
Of course, more lines would go into the first switch...

Also maybe add a constant for the deterministic bit 0x80 in the KDSA function code.